src/Homie/Uptime.cpp

```diff
@@ -4,15 +4,21 @@ using namespace HomieInternals;
 
 Uptime::Uptime()
 : _seconds(0)
-, _lastTick(0) {
+, _lastTick(0)
+, _rolloverCount(0)   {
 }
 
 void Uptime::update() {
   uint32_t now = millis();
-  _seconds += (now - _lastTick) / 1000UL;
+  
+  if (now < _lastTick) {
+    _rolloverCount++;
+  }
+  
   _lastTick = now;
+  _seconds =  (0xFFFFFFFF / 1000 ) * _rolloverCount + (_lastTick / 1000);
```

Comment:
Actually, it might be worse than that. `0xFFFFFFFF / 1000` is integer division, giving the result `4294967`, which is missing the fractional part of 0.295. So it loses 295 milliseconds each rollover (actually 296, as above).

It's OK to round down to the nearest integer number of seconds overall since that's the objective, but I think it's probably not OK to be out by more than 1 second.